components/org.wso2.carbon.identity.application.authenticator.google/src/main/java/org/wso2/carbon/identity/application/authenticator/google/GoogleOAuth2Authenticator.java

```diff
@@ -58,6 +70,158 @@ private void initTokenEndpoint() {
+    public boolean canHandle(HttpServletRequest request) {
+
+        boolean canHandle = false;
+        if (log.isDebugEnabled()) {
+            log.debug("Inside OpenIDConnectAuthenticator.canHandle()");
+        }
+        if (StringUtils.isNotBlank(request.getParameter("oneTapEnabled")) &&
+                request.getParameter("oneTapEnabled").equals("YES")) {
+            // Verifying the Cross-Site Request Forgery (CSRF) token.
+            boolean validCookies = validateCSRFCookies(request);
+            if (validCookies) {
+                canHandle = validateJWTFromGOT(request);
+            }
+        } else {
+            canHandle = super.canHandle(request);
+        }
+        return canHandle;
+    }
+
+    /**
+     * This function validated the JWT token sent via Google One Tap respond
+     * @param request
+     * @return
+     */
+    private boolean validateJWTFromGOT(HttpServletRequest request) {
+
+        boolean validCredential;
+        ExternalIdPConfig externalIdPConfig;
+        try {
+            externalIdPConfig = ConfigurationFacade.getInstance()
+                    .getIdPConfigByName("Google Authenticator", "carbon.super");
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        String clientId = FrameworkUtils
+                .getAuthenticatorPropertyMapFromIdP(externalIdPConfig, "GoogleOIDCAuthenticator").get("ClientId");
+
+        String idTokenString = request.getParameter("credential");
+        //Verifying the ID token.
+        ApacheHttpTransport transport = new ApacheHttpTransport();
+        GsonFactory jsonFactory = new GsonFactory();
+
+        // Specify the CLIENT_ID of the app that accesses the backend:
+        GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(transport, jsonFactory)
+                .setAudience(Collections.singletonList(clientId))
+                // Or, if multiple clients access the backend:
+                //.setAudience(Arrays.asList(CLIENT_ID_1, CLIENT_ID_2, CLIENT_ID_3))
+                .build();
+        GoogleIdToken idToken;
+        try {
+             idToken = verifier.verify(idTokenString);
+        } catch (GeneralSecurityException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("In-secured JWT returned from Google One Tap", e);
+            }
+            throw new RuntimeException(e);
+        } catch (IOException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Exception while validating the JWT returned from Google One Tap");
+            }
+            throw new RuntimeException(e);
+        }
+        validCredential =  idToken != null ? true:false;
+        return validCredential;
+    }
+
+    /**
+     * This function validates the CSRF double-sided cookie returned from Google One Tap respond
+     * The request is considered as non-attacked request if the CSRF cookie and the parameter is equal
+     * @param request
+     * @return
+     */
```

Comment:
Shall we fix this comment properly?

Reply:
Fixed f2aa0c1666f9b5234e7d113125fadd55008081cd